src/driver/hermes_cdev.c

```diff
@@ -130,6 +141,12 @@ static int hermes_fsync(struct file *filp, loff_t start, loff_t end, int datasyn
 		return -EBADFD;
 	}
 
+	eng = hermes_get_ebpf_eng(hermes);
+	if (eng == -ENOSPC)
+		return -EBUSY;
```

Comment:
Probably better to design something to wait for an engine to be not busy, instead of failing in the rare case that there is no busy engine. Application writers are probably not going to get it right otherwise.

Reply:
Yeah, this isn't a great solution.

Use a wait queue. If there is no free ID,  loop forever waiting on that queue (with an interruptible wait that returns an error if it gets interrupted) until it gets an ID.

Whenever a free ID is returned to the ida, wake one sleeper.